cpctools/lib/libdsk/man/dsktrans.1
.\" -*- nroff -*-
.\"
.\" dsktrans.1: dsktrans man page
.\" Copyright (c) 2002 Darren Salt
.\" Copyright (c) 2005-6 John Elliott
.\"
.\" This library is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU Library General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or (at
.\" your option) any later version.
.\"
.\" This library is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU Library General Public License
.\" along with this library; if not, write to the Free Software Foundation,
.\" Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
.\"
.\" Author contact information:
.\" John Elliott: email: jce@seasip.demon.co.uk
.\"
.TH dsktrans 1 "11 April, 2006" "Version 1.1.10" "Emulators"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
dsktrans - Copy from one floppy or image file to another
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.PD 0
.B dsktrans
.RI [ "-itype TYPE" ]
.RI [ "-otype TYPE" ]
.RI [ "-iside SIDE" ]
.RI [ "-oside SIDE" ]
.RI [ "-icomp COMP" ]
.RI [ "-ocomp COMP" ]
.RI [ "-idstep" ]
.RI [ "-odstep" ]
.RI [ "-retry COUNT" ]
.RI [ "-format FMT" ]
.RI [ "-first CYLINDER" ]
.RI [ "-last CYLINDER" ]
.RI [ -md3 ]
.RI [ -logical ]
.RI [ -apricot ]
.RI [ -pcdos ]
.RI [ -noformat ]
.I INPUT-IMAGE
.I OUTPUT-IMAGE
.P
.PD 1
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
Dsktrans copies floppy discs or images, optionally converting image types.
This is a high-level copy which assumes that the disc has a straightforward
geometry; all the tracks are assumed to have the same layout of sectors,
and interleave is not preserved. See
.I dskdump(1)
for a slower but more accurate copy which may preserve more of these 
details.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.B -itype TYPE
Determines which driver is to be used to read from the source disc.
.RS
.TP 8
.B auto
Select according to the disc image file. This is the default.
.TP
.B dsk
Use the DSK (CPCEmu format) image driver.
.TP
.B edsk
Use the extended version of the DSK format.
.TP
.B floppy
Use the floppy driver.
.TP
.B myz80
Use the hard disk (MYZ80 format) image driver.
(This format cannot be autodetected.)
.TP
.B cfi
Use the CFI (DOS fdcopy format) image driver.
(This format cannot be autodetected.)
.TP
.B apridisk
Use the ApriDisk image driver (from the utility of the same name).
(This format cannot be autodetected.)
.TP
.B raw
Use the raw driver.
.RE

.TP
.B -otype TYPE
Determines which driver is to be used to write to the destination disc. The
drivers are as for -itype.

.TP
.B -icomp COMP
Select the compression method used on the source disc image file (has no
effect when reading a floppy disc).
.RS
.TP 8
.B auto
Detect from the first few bytes of the file. This is the default.
.TP
.B sq
Huffman coded (SQ / USQ).
.TP
.B gz
Gzipped (gzip / gunzip).
.TP
.B bz2
Burrows-Wheeler compressed (bzip2 / bunzip2).
.RE

.TP
.B -ocomp COMP
Select the compression to be used on output. Compression methods are as for
-icomp, except that bz2 cannot be used.

.TP
.B -iside SIDE
Determines which side (0 or 1) of the source disc is to be read from.

.TP
.B -oside SIDE
Determines which side (0 or 1) of the destination disc is to be written to.

.TP
.B -idstep
Double-step the source drive (used to read 360k discs in 1.2Mb drives). Only
supported by the Linux floppy driver.

.TP
.B -odstep
Double-step the destination drive (used to write 360k discs in 1.2Mb drives).
Only supported by the Linux floppy driver.

.TP
.B -retry COUNT
Set the number of times to attempt a read/write/format in case of error.

.TP
.B -format FMT
Do not autodetect the disc format; use the named format.

.TP
.B -first CYL 
Start copying at the specified cylinder. Cylinders prior to this will not be 
formatted or written.

.TP
.B -last CYL 
Copy up to and including the specified cylinder.

.TP
.B -odstep
Double-step the destination drive (used to write 360k discs in 1.2Mb drives).
Only supported by the Linux floppy driver.

.TP
.B -md3
Defeat MicroDesign 3 copy protection. Note that this does not make
.I dsktrans
a circumvention device, since the authors of MicroDesign have placed it in
the public domain and given permission for the copy-protection to be reverse
engineered; I posted their original press release to USENET as 
<1008359853.26849.0.nnrp-13.c2de7091@news.demon.co.uk>.

.TP 
.B -pcdos
Convert the first sector from an Apricot superblock to a PC-DOS superblock.
This allows Apricot-format discs to be imaged as files (with the output 
image type as raw) and then loopback-mounted under Linux.

.TP 
.B -apricot
Reverse -pcdos, and convert the first sector from a PC-DOS superblock to 
an Apricot superblock. 
.I "Note that this is the opposite of what this option did in LibDsk 1.1.9 and earlier."

.TP
.B -logical
Rearrange the tracks in the logical order. This option is useful when the
output image type is raw and you're reading a disc with a non-IBM track
layout, such as 144FEAT 1440K or ADFS 'L'.

.TP
.B -noformat
Don't format the target disc/image - assume it's in the correct format
already.
.\"
.\"------------------------------------------------------------------
.\"
.\".SH BUGS
.\"
.\"------------------------------------------------------------------
.\"
.SH SEE ALSO
dskdump(1)
.\"
.\"------------------------------------------------------------------
.\"
.\" `AUTHOR' here is deliberate...
.\"
.SH AUTHOR
John Elliott <jce@seasip.demon.co.uk>.
.PP
Darren Salt wrote the man pages.


cpctools/NEWS
2008/09/17:
	Now works with gcc-4.3


cpctools/lib/libdsk/win32vc6/dsktrans.c
/***************************************************************************
 *                                                                         *
 *    LIBDSK: General floppy and diskimage access library                  *
 *    Copyright (C) 2001-2,2005  John Elliott <jce@seasip.demon.co.uk>     *
 *                                                                         *
 *    This library is free software; you can redistribute it and/or        *
 *    modify it under the terms of the GNU Library General Public          *
 *    License as published by the Free Software Foundation; either         *
 *    version 2 of the License, or (at your option) any later version.     *
 *                                                                         *
 *    This library is distributed in the hope that it will be useful,      *
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of       *
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU    *
 *    Library General Public License for more details.                     *
 *                                                                         *
 *    You should have received a copy of the GNU Library General Public    *
 *    License along with this library; if not, write to the Free           *
 *    Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,      *
 *    MA 02111-1307, USA                                                   *
 *                                                                         *
 ***************************************************************************/

/* Portable equivalent of PCWTRANS */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include "libdsk.h"
#include "utilopts.h"
#include "formname.h"

#ifdef CPM
#define AV0 "DSKTRANS"
#else
#define AV0 argv[0]
#endif

#define DSKTRANS_DEBUG 0

static int md3 = 0;
static int apricot = 0;
static int logical = 0;
static int noformat = 0;
static dsk_format_t format = -1;
static	char *intyp = NULL, *outtyp = NULL;
static	char *incomp = NULL, *outcomp = NULL;
static	int inside = -1, outside = -1;
static  int idstep =  0, odstep  =  0;
static  int retries = 1;
static int first = -1, last = -1;

int do_copy(char *infile, char *outfile);

int check_numeric(char *arg, int *argc, char **argv)
{
	int n = find_arg(arg, *argc, argv);
	unsigned nr;

	if (n < 0) return -1;	
	excise_arg(n, argc, argv);
	if (n >= *argc || atoi(argv[n]) == 0)
	{
		fprintf(stderr, "Syntax error: use '%s nnn' where nnn is nonzero\n", arg);
		exit(1);
	}
	nr = atoi(argv[n]);
	excise_arg(n, argc, argv);

	return nr;
}


static void report(const char *s)
{
        printf("%s\r", s);
        fflush(stdout);
}

static void report_end(void)
{
        printf("\r%-79.79s\r", "");
        fflush(stdout);
}


int help(int argc, char **argv)
{
	fprintf(stderr, "Syntax: \n"
                       "      %s {options} in-image out-image\n",
			AV0);
	fprintf(stderr,"\nOptions are:\n"
		       "-itype <type>   type of input disc image\n"
                       "-otype <type>   type of output disc image\n"
                       "-iside <side>   Force side 0 or side 1 of input\n"
                       "-oside <side>   Force side 0 or side 1 of output\n"
                       "-first <cyl>    Copy starting from specified cylinder\n"
                       "-last <cyl>     Copy up to and including specified cylinder\n"
		       "-retry <count>  Set number of retries on error\n"
		       "-idstep         Double-step when reading\n"
		       "-odstep         Double-step when writing\n"
                       "-noformat       Do not format destination disc\n"
                       "-md3            Defeat MicroDesign 3 copy protection\n"
                       "-apricot        Convert Apricot superblock to PC-DOS format\n"
		       "-logical        Rearrange tracks in logical order\n"
                       "                (Only useful when out-image type is 'raw' and reading discs\n"
		       "                with non-IBM track layout (eg: 144FEAT 1.4Mb or ADFS 640k)\n"
		       "-format         Force a specified format name\n");
	fprintf(stderr,"\nDefault in-image type is autodetect."
		               "\nDefault out-image type is DSK.\n\n");
		
	fprintf(stderr, "eg: %s /dev/fd0 myfile1.DSK\n"
                        "    %s -side 1 /dev/fd0 myfile2.DSK\n" 
                        "    %s -md3 /dev/fd0 md3boot.dsk\n" 
                        "    %s -otype floppy myfile.DSK /dev/fd0\n", 
			AV0, AV0, AV0, AV0);
	valid_formats();
	return 1;
}


/* A quick-and-dirty converter which replaces an Apricot MS-DOS 
 * boot block with a standard PCDOS one. Intended use is for 
 * loopback mounting:
 *
 * dsktrans -otype raw -apricot /dev/fd0 filename.ufi
 * mount -o loop -t msdos filename.ufi /mnt/floppy
 */

void apricot_bootsect(unsigned char *bootsect) 
{
	int n;
	int heads     = bootsect[0x16];
	int sectors   = bootsect[0x10] + 256 * bootsect[0x11];
	char label[8];

/* Check that the first 8 bytes are ASCII or all zeroes. */
	for (n = 0; n < 8; n++)
	{
		if (bootsect[n] != 0 && 
			(bootsect[n] < 0x20 || bootsect[n] > 0x7E)) return;
	}
	memcpy(label, bootsect, 8);
	bootsect[0] = 0xE9;	/* 80x86 jump */
	bootsect[1] = 0x40;
	bootsect[2] = 0x90;
	memcpy(bootsect + 3, label, 8);	/* OEM ID */
	memcpy(bootsect + 11, bootsect + 80, 13);	/* BPB */
	bootsect[24] = sectors & 0xFF;
	bootsect[25] = sectors >> 8;
	bootsect[26] = heads & 0xFF;
	bootsect[27] = heads >> 8;
	memset(bootsect + 28, 0, 512 - 28);
	bootsect[0x40] = 0x90;	/* Minimal boot code: INT 18, diskless boot */
	bootsect[0x41] = 0x90;  /* or ROM BASIC (delete as applicable) */
	bootsect[0x42] = 0xcd;
	bootsect[0x43] = 0x18;
}



int main(int argc, char **argv)
{

	if (find_arg("--version", argc, argv) > 0) return version(); 
	if (argc < 3) return help(argc, argv);
	if (find_arg("--help",    argc, argv) > 0) return help(argc, argv);

	ignore_arg("-type", 2, &argc, argv);
	ignore_arg("-side", 2, &argc, argv);
	ignore_arg("-dstep", 2, &argc, argv);
	ignore_arg("-comp", 2, &argc, argv);

	intyp     = check_type("-itype", &argc, argv); 
        outtyp    = check_type("-otype", &argc, argv);
	incomp    = check_type("-icomp", &argc, argv); 
        outcomp   = check_type("-ocomp", &argc, argv);
        inside    = check_forcehead("-iside", &argc, argv);
        outside   = check_forcehead("-oside", &argc, argv);
	retries   = check_retry("-retry", &argc, argv);
	first     = check_numeric("-first", &argc, argv);
	last      = check_numeric("-last", &argc, argv);
	if (present_arg("-idstep", &argc, argv)) idstep = 1;
	if (present_arg("-odstep", &argc, argv)) odstep = 1;
	if (present_arg("-md3", &argc, argv)) md3 = 1;
	if (present_arg("-apricot", &argc, argv)) apricot = 1;
	if (present_arg("-noformat", &argc, argv)) noformat = 1;
	if (present_arg("-logical", &argc, argv)) logical = 1;
	if (!outtyp) outtyp = "dsk";
        format    = check_format("-format", &argc, argv);
	args_complete(&argc, argv);
	return do_copy(argv[1], argv[2]);
}



int do_copy(char *infile, char *outfile)
{
	DSK_PDRIVER indr = NULL, outdr = NULL;
	dsk_err_t e;
	dsk_pcyl_t cyl;
	dsk_phead_t head;
	dsk_psect_t sec;
	char *buf = NULL;
	char *cmt = NULL;
	DSK_GEOMETRY dg;
	char *op = "Opening";

        dsk_reportfunc_set(report, report_end);

	        e = dsk_open (&indr,  infile,  intyp, incomp);
	if (!e) e = dsk_set_retry(indr, retries);
	if (!e && inside >= 0) e = dsk_set_option(indr, "HEAD", inside);
	if (!e) e = dsk_creat(&outdr, outfile, outtyp, outcomp);
	if (!e && outside >= 0) e = dsk_set_option(outdr, "HEAD", outside);
	if (!e && idstep) e = dsk_set_option(indr, "DOUBLESTEP", 1);
	if (!e && odstep) e = dsk_set_option(outdr, "DOUBLESTEP", 1);
	if (!e) e = dsk_set_retry(outdr, retries);
	if (format == -1)
	{
		op = "Identifying disc";
		if (!e) e = dsk_getgeom(indr, &dg);
	}
	else if (!e) e = dg_stdformat(&dg, format, NULL, NULL);
	if (!e)
	{	
		buf = dsk_malloc(dg.dg_secsize);
		if (!buf) e = DSK_ERR_NOMEM;
	}
	if (!e)
	{
		/* Copy comment, if any */
		dsk_get_comment(indr, &cmt);
		dsk_set_comment(outdr, cmt);
		printf("Input driver: %s\nOutput driver:%s\n%s",
			dsk_drvdesc(indr), dsk_drvdesc(outdr),
			logical ? "[tracks rearranged]\n" : "");
		if (first < 0) first = 0;
		if (last < 0) last = dg.dg_cylinders - 1;

		for (cyl = first; cyl <= (dsk_pcyl_t)last; ++cyl)
		{
		    for (head = 0; head < dg.dg_heads; ++head)
		    {
                        if (md3)
                        {
			// MD3 discs have 256-byte sectors in cyl 1 head 1
			    if (cyl == 1 && head == 1) dg.dg_secsize = 256;
			    else		       dg.dg_secsize = 512;
                        }
			// Format track!
                        if (e == DSK_ERR_OK && (!noformat)) 
			{
				op = "Formatting";
				e = dsk_apform(outdr, &dg, cyl, head, 0xE5);
			}
			if (!e) for (sec = 0; sec < dg.dg_sectors; ++sec)
			{
				printf("Cyl %02d/%02d Head %d/%d Sector %03d/%03d\r", 
					cyl +1, dg.dg_cylinders,
				 	head+1, dg.dg_heads,
					sec+dg.dg_secbase, dg.dg_sectors + dg.dg_secbase - 1); 
				fflush(stdout);
			
				op = "Reading";	
				if (logical)
				{
					dsk_lsect_t ls;
					dsk_sides_t si;

	/* Convert sector to logical using SIDES_ALT sidedness. Raw DSKs 
	 * are always created so that the tracks are stored in SIDES_ALT 
	 * order. */
					si = dg.dg_sidedness;
					dg.dg_sidedness = SIDES_ALT;
					dg_ps2ls(&dg, cyl, head, sec, &ls);
					dg.dg_sidedness = si;
					e = dsk_lread(indr, &dg, buf, ls);
				}
				else e = dsk_pread(indr, &dg, buf, cyl,head,sec + dg.dg_secbase);
				// MD3 discs have deliberate bad sectors in cyl 1 head 1
                                if (md3 && e == DSK_ERR_DATAERR && dg.dg_secsize == 256) e = DSK_ERR_OK;
				if (e) break;
				op = "Writing";
#if DSKTRANS_DEBUG
{ 
 int xn, yn;
 char xbuf[20];
 for (xn = 0; xn < sizeof(xbuf)/2; xn++)
 {
  if (isprint(buf[xn])) xbuf[xn] = buf[xn]; else xbuf[xn] = '.';
 }
 yn = dg.dg_secsize - (sizeof(xbuf)-1) + xn;
 for (; xn < sizeof(xbuf)-1; xn++, yn++)
 {
  if (isprint(buf[yn])) xbuf[xn] = buf[yn]; else xbuf[xn] = '.';
 }
 xbuf[xn] = 0;

 printf("WR: c%02d h%d s%d %02x %02x %02x %02x %s\n", 
         cyl, head, sec, 
	 buf[0] & 0xFF, buf[1] & 0xFF, buf[2] & 0xFF, buf[3] & 0xFF, xbuf);
}
#endif
				if (apricot && cyl == 0 && head == 0 && 
						sec == 0)
					apricot_bootsect((unsigned char *)buf);
				e = dsk_pwrite(outdr,&dg,buf,cyl,head, sec + dg.dg_secbase);
				if (e) break;
			}
			if (e) break;
		    }
		    if (e) break;
		}
	
	}
	printf("\r                                     \r");
	if (indr)  dsk_close(&indr);
	if (outdr) dsk_close(&outdr);
	if (buf) dsk_free(buf);
	if (e)
	{
		fprintf(stderr, "\n%s: %s\n", op, dsk_strerror(e));
		return 1;
	}
	return 0;
}

